### DISCUSSION TITLE
您好，请教一个关于 Docker 中使用非 root 用户运行后的 VOLUME 的权限问题

### DISCUSSION THREAD

**Discussion by GoGoogle** [gythialy/gythialy.github.io#142] [General]
看了您的[文章](https://github.com/gythialy/gythialy.github.io/blob/raw/source/_posts/non-root-user-inside-a-Docker-container.md)中的dockerfile最后一句是：
`VOLUME ["$QLCHOME"]`
也就是
`VOLUME /qlcchain`

那当这个镜像构建成功后，在使用的它的时候：

`docker run -d -t --name test -v /my:/qlcchain thiscontainerimages /usr/local/bin/gqlc --version>/qlcchain/test.txt`

(假如说 `/usr/local/bin/gqlc --version` 可以显示版本，并写入到 `/qlcchain/test.txt` ）

这样一来，我宿主机下应该会有 `/my/test.txt` 吧？

但经过我的测试发现没有权限写入 `/qlcchain/test.txt`

所以，这时候应该如何解决这种问题？

不胜感激！

**Comment by gythialy** [comment]
在我的示例中，在 alpine 的系统下建出来的用户，`uid=100,gid=1000`，所以你需要把你挂载在容器内的目录增加容器内用户的读写权限， `chown -R 100:1000 my`

**Comment by GoGoogle** [comment]
> 在我的示例中，在 alpine 的系统下建出来的用户，`uid=100,gid=1000`，所以你需要把你挂载在容器内的目录增加容器内用户的读写权限， `chown -R 100:1000 my`

这样uid还是跟宿主机不一样，也会有问题，我后来用su-exec替代了直接运行程序。

**Comment by gythialy** [comment]
不需要和宿主机一样，只要挂载到容器内的目录，在容器内的uid用户有权限即可。不过 su-exec/gosu 这种也是一个解决办法

**Comment by GoGoogle** [comment]
> 不需要和宿主机一样，只要挂载到容器内的目录，在容器内的uid用户有权限即可。不过 su-exec/gosu 这种也是一个解决办法

与宿主机共享数据的时候，只有容器uid用户权限就不合适，无法修改。
发现gosu需要另外wget，alpine里面有su-exec 所以，最终就选择了 su-exec
